nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/ControllerEndpointMerger.java

```diff
@@ -0,0 +1,113 @@
+        return entity.getController();
+    }
+
+    @Override
+    protected void mergeResponses(ControllerDTO clientDto, Map<NodeIdentifier, ControllerDTO> dtoMap, Set<NodeResponse> successfulResponses, Set<NodeResponse> problematicResponses) {
+        ControllerDTO mergedController = clientDto;
+        final Map<String, Map<NodeIdentifier, PortDTO>> inputPortMap = new HashMap<>(); // map of port id to map of node id to port dto
+        final Map<String, Map<NodeIdentifier, PortDTO>> outputPortMap = new HashMap<>(); // map of port id to map of node id to port dto
+
+        for (final Map.Entry<NodeIdentifier, ControllerDTO> entry : dtoMap.entrySet()) {
+            final NodeIdentifier nodeId = entry.getKey();
+            final ControllerDTO nodeController = entry.getValue();
+
+            if (nodeController == mergedController) {
+                continue;
+            }
+
+            nodeController.getInputPorts().stream().forEach(inputPort -> inputPortMap.computeIfAbsent(inputPort.getId(), nodeIdToInputPort -> new HashMap<>()).put(nodeId, inputPort));
+            nodeController.getOutputPorts().stream().forEach(outputPort -> outputPortMap.computeIfAbsent(outputPort.getId(), nodeIdToOutputPort -> new HashMap<>()).put(nodeId, outputPort));
+        }
+
+        /*
+         * Note on port merging: only merge the ports if they exist in the client response and all node responses.  Due to authorization possibly different per node, only ports that have been
+         * returned from every node need to be merged.  If a node doesn't return a port DTO due to authorizatino issues, the responses for that port ID should be dropped from the client response.
+         */
+
+        // merge input ports
+        for (Map<NodeIdentifier, PortDTO> inputPortByNodeId : inputPortMap.values()) {
+            final Collection<PortDTO> nodeInputPorts = inputPortByNodeId.values();
+            if (!nodeInputPorts.isEmpty()) {
+                final PortDTO inputPort = nodeInputPorts.iterator().next();
+                final PortDTO clientInputPort = clientDto.getInputPorts().stream().filter(p -> p.getId().equals(inputPort.getId())).findFirst().orElse(null);
+                if (clientInputPort != null) {
+                    PortEntityMerger.mergeDtos(clientInputPort, inputPortByNodeId);
+                }
+            }
+        }
+
+        // merge output ports
+        for (Map<NodeIdentifier, PortDTO> outputPortByNodeId : outputPortMap.values()) {
+            final Collection<PortDTO> nodeOutputPorts = outputPortByNodeId.values();
+            if (!nodeOutputPorts.isEmpty()) {
+                final PortDTO outputPort = nodeOutputPorts.iterator().next();
+                final PortDTO clientOutputPort = clientDto.getInputPorts().stream().filter(p -> p.getId().equals(outputPort.getId())).findFirst().orElse(null);
+                if (clientOutputPort != null) {
+                    PortEntityMerger.mergeDtos(clientOutputPort, outputPortByNodeId);
```

Comment:
It looks like `PortEntityMerger.mergeDtos` is expecting the `dtoMap` (2nd argument) to contain all DTOs including the `clientDto`. In the `ControllerEndpointMerger` it looks like the `dtoMap` will contain all DTOs except the `clientDto`. As a result, it looks like any `validationErrors` in `clientDto` will be lost.